### DISCUSSION TITLE
Allow selectinload to skip join and filter directly on the target table (2)

### DISCUSSION THREAD

**Discussion by nmoreaud** [sqlalchemy/sqlalchemy#10145] [Usage Questions] [answered]
Hello

There is an existing discussion on the subject : https://github.com/sqlalchemy/sqlalchemy/issues/4340
I wonder why the "many to one" case could not be optimized the same way.

Let's suppose I have a db with tables like :
User
  - id
  - country_code
  - firstname
  - lastname

Country
  - id
  - code    (key)
  - name
  - epsg
  - language

```
query(User).option(selectinload(User.country)).limit(5000).all()
```

I wonder if the ORM could directly emit:

```
select * from country where code in ('us', 'fr', 'pl');
```

rather than:

```
select * from user
join country on country.code = user.country_code
where uiser.id in (1, 2, 4, 5, 8, 9, 3, 4, ...);
```

I have a lot more users in my database than countries ; it would reduce drastically the `where in` condition.
It could be based on an option.

There also could exist a load option that loads the full country table without `where`, because I know it is small.

**Comment by zzzeek** [comment]
hi - 

many-to-one is already optimized to not use any JOIN.  it selects from the many-to-one primary key value extracted from the parent objects and only queries the related table directly.  so if you have lots of primary rows that all link to a set of five many-to-ones, the IN will have only five identifiers in it.   this was implemented by #4775 for SQLAlchemy 1.3.6.

**Comment by nmoreaud** [comment]
Nice. it doesn't seem to work with my program, I will investigate why

**Reply by zzzeek** [reply]
It would be because you are using a custom primaryjoin condition in your many to one, so that it is not a pure m2o.

**Reply by nmoreaud** [reply]
My models are defined like this:
```python
class User(Base):
    __tablename__ = 'users'

    id = Column(Integer, primary_key=True)
    firstname = Column(String(255))
    country_code = Column(ForeignKey('countries.code', ondelete='RESTRICT', onupdate='CASCADE'))

    country = relationship('Country')
    

class Country(Base):
    __tablename__ = 'countries'

    id = Column(Integer, primary_key=True)
    code = Column(String(50), nullable=False, unique=True)
```

**Reply by nmoreaud** [reply]
![image](https://github.com/sqlalchemy/sqlalchemy/assets/2413781/06a4672a-b2f9-45cd-b768-62864d0e829f)
use_get is False

**Reply by nmoreaud** [reply]
![image](https://github.com/sqlalchemy/sqlalchemy/assets/2413781/89fcc2c9-9e02-4be7-8d66-851cfca763f7)
`compare` returns false but I cannot enter it, do you know where is the implementation?
Edit: found it

**Reply by nmoreaud** [reply]
Okay, I guess that joining on a unique key is not supported. It has to be the primary key?

**Reply by CaselIT** [reply]
> Okay, I guess that joining on a unique key is not supported. It has to be the primary key?

That's likely the cause. @zzzeek do you think it's worth adding support for unique keys here?

**Reply by zzzeek** [reply, marked as answer]
while selectinload() could get away with this, use_get in general would not work because the ORM has to load objects by primary key.

So there's a super easy way to set this up which is to tell the ORM that "code" is the primary key.    No table changes are necessary.  Try it out!

```py
import random

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import Session

Base = declarative_base()

class User(Base):
    __tablename__ = 'users'

    id = Column(Integer, primary_key=True)
    firstname = Column(String(255))
    country_code = Column(ForeignKey('countries.code', ondelete='RESTRICT', onupdate='CASCADE'))

    country = relationship('Country')


class Country(Base):
    __tablename__ = 'countries'

    id = Column(Integer, primary_key=True)
    code = Column(String(50), nullable=False, unique=True)

    __mapper_args__ = {
        "primary_key": code
    }

e = create_engine("sqlite://", echo=True)


Base.metadata.create_all(e)

with Session(e) as sess:
    US, JP, FR = Country(code="US"), Country(code="JP"), Country(code="FR")

    sess.add_all([
        User(firstname=f"user {i}", country=random.choice([US, JP, FR])) for i in range(50)
    ])

    sess.commit()

with Session(e) as sess:
    sess.execute(select(User).options(selectinload(User.country))).all()
```

in the output there's your SQL

```
2023-07-26 14:10:03,619 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-07-26 14:10:03,621 INFO sqlalchemy.engine.Engine SELECT users.id, users.firstname, users.country_code 
FROM users
2023-07-26 14:10:03,621 INFO sqlalchemy.engine.Engine [generated in 0.00012s] ()
2023-07-26 14:10:03,623 INFO sqlalchemy.engine.Engine SELECT countries.code AS countries_code, countries.id AS countries_id 
FROM countries 
WHERE countries.code IN (?, ?, ?)
2023-07-26 14:10:03,623 INFO sqlalchemy.engine.Engine [generated in 0.00016s] ('FR', 'JP', 'US')
2023-07-26 14:10:03,624 INFO sqlalchemy.engine.Engine ROLLBACK

```

**Reply by zzzeek** [reply]
this will also help dramatically with your lazyloads etc. since they will pull from identity map

**Reply by nmoreaud** [reply]
This is a nice workaround, thanks! It works well for my use case.
It would be interesting to handle unique keys by default thought.

**Reply by nmoreaud** [reply]
It seems to have other implications, it changes the behavior of `session.get(Country, id)`

**Reply by CaselIT** [reply]
Well yes, code should now be used as the primary key for a country, so `session.get(Country, code)`
